feat: Add support for attaching namespace metadata to Kubernetes SD

dongjiang1989 opened the pull request:
## Description

This PR adds support for attaching namespace metadata to Kubernetes SD discovered targets.

When `attachMetadata.namespace` is set to `true`, Prometheus attaches namespace metadata to the discovered targets. This requires Prometheus >= v3.6.0.

The changes include:
- Add `Namespace` field to `AttachMetadata` struct in both `v1` and `v1alpha1` APIs
- Update `promcfg` to generate `attach_metadata.namespace` in scrape config for supported roles
- Add CEL validation rules to `KubernetesSDConfig` in ScrapeConfig CRD:
  - `attachMetadata.node` is only valid for `Pod`, `Endpoints`, and `EndpointSlice` roles
  - `attachMetadata.namespace` is only valid for `Pod`, `Endpoints`, `EndpointSlice`, `Service`, and `Ingress` roles
- Update `mergeAttachMetadataForTopology` to preserve namespace setting when topology sharding forces node=true
- Add conversion helper for `v1alpha1.AttachMetadata` to `v1.AttachMetadata`
- Update generated CRDs, apply configurations, and golden test files

Closes: #7799

## Type of change

- [x] `FEATURE` (non-breaking change which adds functionality)

## Verification

- All existing unit tests pass (`go test ./pkg/...`)
- Added new test cases for namespace metadata attachment in `promcfg_test.go`
- Added golden test files for various scenarios:
  - Supported role with namespace metadata
  - Unsupported role with namespace metadata
  - Unsupported version with namespace metadata
- E2E test updated to use `v1alpha1.AttachMetadata` type

## Changelog entry

```release-note
Add support for attaching namespace metadata to Kubernetes SD
```

simonpasquier reviewed (commented):
Thanks for the PR! Don't we have an existing issue with ScrapeConfig CRD and scrape classes? It seems that we don't merge the scrape class' attach metadata when generating the configuration from ScrapeConfig.

dongjiang1989 commented:
> Thanks for the PR! Don't we have an existing issue with ScrapeConfig CRD and scrape classes? It seems that we don't merge the scrape class' attach metadata when generating the configuration from ScrapeConfig.

Thanks @simonpasquier 
Currently, there appears to be no issue, as the results have been uniformly processed via `appendAttachMetadata`.

sebastienrospars commented:
Thanks for the work on this @dongjiang1989 ! 🙏 
Is this PR still active? I’d love to use the namespace metadata feature and was wondering about its status.

dongjiang1989 commented:
@heliapb and @sebastienrospars  I'll fix it soon

dongjiang1989 commented:
@simonpasquier @heliapb  PTAL

sebastienrospars commented:
Hello @heliapb, do you have some times to review this PR please ? Thanks by advance.

simonpasquier commented on pkg/apis/monitoring/v1/types.go:
```diff
@@ -705,6 +705,19 @@ type AttachMetadata struct {
+	// The Prometheus service account must have the `list` and `watch`
+	// permissions on the `Namespaces` objects.
+	//
+	// Only valid for pod, endpoints, endpointslice, service, ingress.
```
we don't need the comment since it's "only" supported for ServiceMonitor/PodMonitor/Probe?

simonpasquier commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -466,7 +475,7 @@ type KubernetesSDConfig struct {
 	// Prometheus >= v2.37.0 for `Endpoints` and `Endpointslice` roles.
 	// +optional
-	AttachMetadata *AttachMetadata `json:"attachMetadata,omitempty"`
+	AttachMetadata *v1.AttachMetadata `json:"attachMetadata,omitempty"`
```
Why switch to the v1 struct if we're also modifying the v1alpha1 struct?
I'd prefer to keep the current approach of duplicating the structs, especially since the restrictions are a bit different.

dongjiang1989 commented:
@simonpasquier @slashpai   Rebase done. Please re-check it

/cc @aivanzipper  @sebastienrospars

simonpasquier commented on pkg/apis/monitoring/v1alpha1/scrapeconfig_types.go:
```diff
@@ -51,12 +51,21 @@ type NamespaceDiscovery struct {
+
+	// namespace attaches namespace metadata to discovered targets.
+	// When set to true, Prometheus must have permissions to list/watch Namespaces.
+	// Only valid for pod, endpoints, endpointslice, service, ingress.
```
should we have CEL validations for this?

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -2351,11 +2351,15 @@ func (cg *ConfigGenerator) getNamespacesFromNamespaceSelector(nsel monitoringv1.
 type attachMetadataConfig struct {
 	MinimumVersion string
-	attachMetadata *monitoringv1.AttachMetadata
+	AttachMetadata *monitoringv1.AttachMetadata
```
(nit) why the change?

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -5332,7 +5351,7 @@ func (cg *ConfigGenerator) mergeAttachMetadataForTopology(amc *attachMetadataCon
 	return &attachMetadataConfig{
 		MinimumVersion: minimumVersion,
-		attachMetadata: &monitoringv1.AttachMetadata{
+		AttachMetadata: &monitoringv1.AttachMetadata{
```
what if namespace == true?

simonpasquier commented on pkg/prometheus/promcfg.go:
```diff
@@ -3569,9 +3586,11 @@ func (cg *ConfigGenerator) generateScrapeConfig(
 				case "endpoints", "endpointslice":
-					configs[i] = cg.WithMinimumVersion("2.37.0").AppendMapItem(configs[i], "attach_metadata", config.AttachMetadata)
+					configs[i] = cg.appendAttachMetadata(configs[i], config.AttachMetadata, "2.37.0")
+				case "service", "ingress":
```
not sure why we add this?

sebastienrospars commented:
Hi @dongjiang1989, do you have some time to take a look at Simon’s feedback when you get a chance? Thanks!

dongjiang1989 committed:
feat: add support for attaching namespace metadata to Kubernetes SD

Add support for attaching namespace metadata to Kubernetes SD in
ServiceMonitor, PodMonitor, Probe and ScrapeConfig CRDs.

- Add Namespace field to AttachMetadata struct (v1 and v1alpha1)
- Update promcfg to generate attach_metadata.namespace in scrape config
- Add CEL validations for KubernetesSDConfig attachMetadata fields:
  - node: only valid for Pod, Endpoints, EndpointSlice roles
  - namespace: only valid for Pod, Endpoints, EndpointSlice, Service, Ingress roles
- Update golden test files and e2e tests
- Fix mergeAttachMetadataForTopology to preserve namespace setting

Closes: #7799
Signed-off-by: dongjiang <dongjiang1989@126.com>

dongjiang1989 commented:
> Hi @dongjiang1989, do you have some time to take a look at Simon’s feedback when you get a chance? Thanks!

Thanks @sebastienrospars 
All  feedback changed done.

cc @simonpasquier  Please re-check it
